GoogleSignIn/Sources/GIDSignInInternalOptions.m

```diff
@@ -31,13 +31,15 @@ + (instancetype)defaultOptionsWithConfiguration:(nullable GIDConfiguration *)con
                                       loginHint:(nullable NSString *)loginHint
                                   addScopesFlow:(BOOL)addScopesFlow
                                          scopes:(nullable NSArray *)scopes
+                                          nonce:(nullable NSString *)nonce
```

Comment:
Should we add the `nonce` parameter to the `+[GIDSignInInternalOptions silentOptionsWithCompletion:]`? This is used to [restore previous sign ins](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Sources/GIDSignIn.m#L223).

Reply:
@mdmathias I believe that nonce is primarily used during the initial authentication. Restoring previous sign in takes existing `idToken` (and other info) if not expired. If expired, afaict, it takes refresh token and goes to exchange it for accessToken and idToken. This is [documented here afaict](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint) and nonce is not mentioned.

Reply:
My understanding of the nonce is that it is used to [associate a client session with an ID token](https://openid.net/specs/openid-connect-core-1_0.html#IDToken). So, it seems reasonable that restoring a sign in (if it [leads to a token refresh as in here](https://github.com/google/GoogleSignIn-iOS/blob/main/GoogleSignIn/Sources/GIDSignIn.m#L600-L613)) would benefit from the nonce. What do you think?

Reply:
hello @mdmathias token refresh leads to a request to `https://oauth2.googleapis.com/token` from [here](https://github.com/openid/AppAuth-iOS/blob/c89ed571ae140f8eb1142735e6e23d7bb8c34cb2/Sources/AppAuthCore/OIDAuthorizationService.m#L429). 
I previously pointed to section [3.1.3](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint) for this. Section [12.1](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens) describes this situation more accurately and it corresponds to what the request actually looks like (from xcode console):

```
<OIDTokenRequest: 0x600002628060, request: <URL: https://oauth2.googleapis.com/token, HTTPBody: gidenv=ios-sim&client_id=abc.apps.googleusercontent.com&refresh_token=1//xyz&emm_support=1&grant_type=refresh_token&device_os=iOS%2017.4&gpsdk=gid-8.0.0>>
```

[12.2 Successful Refresh Response](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse) specifically states that "If an ID Token is returned as a result of a token refresh request ... it SHOULD NOT have a nonce Claim".

Even if it did, AppAuth [would not validate it](https://github.com/openid/AppAuth-iOS/blob/c89ed571ae140f8eb1142735e6e23d7bb8c34cb2/Sources/AppAuthCore/OIDAuthorizationService.m#L638).

So I still believe no change is needed. Looking forward to your response.

Reply:
Thanks for tracking this down and explaining what you've found. I agree with you. Resolving!